In [1]:
# import required packages
import pandas as pd
import datetime
import numpy as np
import matplotlib.pyplot as plt

# Read in data

In [2]:
# read csv data file
cum_prods = pd.read_csv('prod-by-operated-day.csv')
completions = pd.read_csv('completion.csv')
well_indices = pd.read_csv('well-index.csv')

In [3]:
# check each data variable
cum_prods.head(10)

api  cum_gor_180  cum_gor_30  cum_gor_365  cum_gor_545  \
0  33053032160000     1194.321    1190.631     1205.677     1275.677   
1  33053036170000     1017.419     845.320     1205.925     1316.548   
2  33053050680000      938.340    1730.823          NaN          NaN   
3  33025024960000          NaN    1557.266          NaN          NaN   
4  33053051180000     1532.590         NaN          NaN          NaN   
5  33025010660000      599.998     600.023      599.994      599.997   
6  33025006720000      594.862     648.649      881.078      850.790   
7  33053051550000     1398.075    1309.105     1376.159          NaN   
8  33053043160000     1025.894     548.382     1159.324     1164.761   
9  33053049610000     1111.500     353.623          NaN          NaN   

   cum_gor_730  cum_gor_90  cum_oil_180  cum_oil_30  cum_oil_365    ...      \
0     1329.975    1180.290     169683.0     31936.0     262305.0    ...       
1     1399.867     968.606      93461.0     23067.0     151827.0    ...       
2          NaN    1622.325      91096.0     14392.0          NaN    ...       
3          NaN    1409.905          NaN     15786.0          NaN    ...       
4          NaN    1025.867      89874.0     14136.0          NaN    ...       
5      599.993     600.003     110412.0     26404.0     168315.0    ...       
6      904.512     614.443       4554.0      1443.0      53371.0    ...       
7          NaN    1431.240      60584.0     22604.0      84342.0    ...       
8          NaN     869.687      69861.0     17455.0     104623.0    ...       
9          NaN     629.582     123848.0     18729.0          NaN    ...       

   cum_water_365  cum_water_545  cum_water_730  cum_water_90  cum_wc_180  \
0        46855.0        58148.0        66312.0       28376.0      17.605   
1        27718.0        33426.0        38146.0       16756.0      19.486   
2            NaN            NaN            NaN        3483.0      18.487   
3            NaN            NaN            NaN       18878.0         NaN   
4            NaN            NaN            NaN        2853.0       9.132   
5        77257.0        96106.0       111054.0       39228.0      33.320   
6        19626.0        32912.0        44940.0         518.0      15.369   
7        30789.0            NaN            NaN       17779.0      29.403   
8        56351.0        67945.0            NaN       25561.0      31.355   
9            NaN            NaN            NaN       27121.0      27.957   

   cum_wc_30  cum_wc_365  cum_wc_545  cum_wc_730  cum_wc_90  
0     32.070      15.156      14.625      14.371     21.395  
1     23.302      15.438      15.076      15.188     20.949  
2        NaN         NaN         NaN         NaN     10.609  
3     38.552         NaN         NaN         NaN     37.823  
4      0.049         NaN         NaN         NaN      6.892  
5     47.658      31.460      30.903      30.726     37.037  
6     24.765      26.886      28.110      28.975     13.682  
7     30.291      26.743         NaN         NaN     28.900  
8     46.984      35.006      34.543         NaN     35.933  
9     34.113         NaN         NaN         NaN     30.187  

[10 rows x 25 columns]

In [4]:
completions.head(10)

api  bottom  file_number  fluid_bbl  fluid_bbl_per_ft  \
0  33061029770000   20171        27710      63941                 6   
1  33053065160000   20490        30120      55859                 6   
2  33061032390000   15774        29063      46076                 9   
3  33025026550000   20688        29220      67183                 7   
4  33053064110000   21157        29814      59245                 2   
5  33053064120000   21556        29815      60683                 6   
6  33061027470000   20080        26595     127457                13   
7  33061027490000   19969        26597     128705                13   
8  33053063850000   21121        29725      59978                 6   
9  33053063870000   20566        29727      58438                 6   

   fluid_gal  fluid_gal_per_ft  ft_per_stage  lateral_length  max_treat_press  \
0    2685522               273           280            9806             8239   
1    2346078               260           300            9020             8820   
2    1935192               379           221            5102             8982   
3    2821686               307           262            9190             8327   
4    2488290               124           666           20007             9171   
5    2548686               254           333           10006             8571   
6    5353194               552           201            9681             9789   
7    5405610               559           201            9660             9288   
8    2519076               264           317            9521             8772   
9    2454396               272           300            9016             8741   

   max_treat_rate  propp_lbs  propp_lbs_per_ft  stages    top treatment_date  
0            46.3    3222497               328      35  10365        3/14/15  
1            41.4    3360688               372      30  11470        2/27/15  
2            89.6    1676744               328      23  10672        2/16/15  
3            44.1    4035990               439      35  11498        2/15/15  
4            40.6    3522422               176      30   1150        2/15/15  
5            40.5    3667656               366      30  11550        2/15/15  
6            39.1    9722230              1004      48  10399        2/15/15  
7            28.9    9716840              1005      48  10309        2/11/15  
8            57.2    3622980               380      30  11600         2/9/15  
9            42.1    3364956               373      30  11550         2/9/15

In [5]:
well_indices.head(10)

api  file_no                 Current_Operator  \
0  33011013190000    16064      CONTINENTAL RESOURCES, INC.   
1  33025005960000    16327             MARATHON OIL COMPANY   
2  33007015830000    16431  WHITING OIL AND GAS CORPORATION   
3  33007015840000    16435  WHITING OIL AND GAS CORPORATION   
4  33061005310000    16587              EOG RESOURCES, INC.   
5  33061005330000    16628              EOG RESOURCES, INC.   
6  33061005520000    16728              EOG RESOURCES, INC.   
7  33061005580000    16746              EOG RESOURCES, INC.   
8  33025006570000    16758             MARATHON OIL COMPANY   
9  33025006620000    16764             MARATHON OIL COMPANY   

    Current_Well_Name   Lease_Name Lease_Number  \
0          MPHU  44-9         MPHU         44-9   
1        BENZ  14-21H         BENZ       14-21H   
2          BSMU  1305         BSMU         1305   
3          BSMU  3204         BSMU         3204   
4        ROSS  20-16H         ROSS       20-16H   
5         PAUL  1-19H         PAUL        1-19H   
6         RUUD  1-18H         RUUD        1-18H   
7     PEDERSON  1-33H     PEDERSON        1-33H   
8  SCHOLLMEYER  21-4H  SCHOLLMEYER        21-4H   
9     KNUDSVIG  34-7H     KNUDSVIG        34-7H   

                            Original_Operator  Original_Well_Name   Spud_Date  \
0                 CONTINENTAL RESOURCES, INC.          MPHU  44-9  05/06/2010   
1  PROSPECTIVE INVESTMENT & TRADING CO., LTD.         BENZ  1-21H  06/26/2010   
2             WHITING OIL AND GAS CORPORATION          BSMU  1305  04/10/2010   
3             WHITING OIL AND GAS CORPORATION          BSMU  3203  01/28/2012   
4                         EOG RESOURCES, INC.        EDITH  1-16H  05/03/2010   
5                         EOG RESOURCES, INC.         PAUL  1-19H  02/27/2009   
6                         EOG RESOURCES, INC.         RUUD  1-18H  08/18/2010   
7                         EOG RESOURCES, INC.     PEDERSON  1-33H  02/13/2009   
8                        MARATHON OIL COMPANY  SCHOLLMEYER  21-4H  07/20/2009   
9                        MARATHON OIL COMPANY     KNUDSVIG  34-7H  09/20/2010   

        TD        ...                   Footages           Field_Name  \
0      NaN        ...           426 FSL  315 FEL  MEDICINE POLE HILLS   
1  20210.0        ...           660 FSL  400 FWL               BAILEY   
2   9510.0        ...          906 FSL  1197 FEL            BIG STICK   
3   9505.0        ...         1131 FNL  1776 FWL            BIG STICK   
4  14770.0        ...          300 FNL  1200 FWL                ALGER   
5  14412.0        ...           450 FNL  433 FWL             PARSHALL   
6  13796.0        ...           450 FNL  325 FWL             PARSHALL   
7  14380.0        ...           475 FNL  775 FWL             PARSHALL   
8  19608.0        ...          660 FNL  1980 FWL               BAILEY   
9  19970.0        ...          660 FSL  1375 FEL               BAILEY   

  Produced_Pools             Wellbore   Latitude   Longitude Well_Type  \
0      RED RIVER  HORIZONTAL RE-ENTRY  46.091078 -103.681843        OG   
1         BAKKEN           HORIZONTAL  47.446328 -102.671590        OG   
2        MADISON          DIRECTIONAL  47.112737 -103.418030        OG   
3        MADISON          DIRECTIONAL  47.078109 -103.384544        OG   
4         BAKKEN           HORIZONTAL  48.340815 -102.520857        OG   
5         BAKKEN           HORIZONTAL  48.065260 -102.178736        OG   
6         BAKKEN           HORIZONTAL  47.991234 -102.126972        OG   
7         BAKKEN           HORIZONTAL  48.036189 -102.263650        OG   
8         BAKKEN           HORIZONTAL  47.326732 -102.601309        OG   
9         BAKKEN           HORIZONTAL  47.388383 -102.572353        OG   

  Well_Status       CTB  Well_Status_Date  
0           A  404538.0        08/07/2010  
1           A  116327.0        09/19/2010  
2           A  406270.0        05/26/2010  
3           A  406270.0        04/04/2012  
4           A  116587.0    

In [6]:
# check the different columns available in well_indices
well_indices.columns

Index(['api', 'file_no', 'Current_Operator', 'Current_Well_Name', 'Lease_Name',
       'Lease_Number', 'Original_Operator', 'Original_Well_Name', 'Spud_Date',
       'TD', 'County_Name', 'Township', 'Range', 'Section', 'QQ', 'Footages',
       'Field_Name', 'Produced_Pools', 'Wellbore', 'Latitude', 'Longitude',
       'Well_Type', 'Well_Status', 'CTB', 'Well_Status_Date'],
      dtype='object')

# Production data

In [7]:
# isolate the production variable desired
prods_365 = cum_prods.set_index('api')[['cum_oil_365']]

In [8]:
# check how many nan values
prods_365.isnull().sum()

cum_oil_365    2480
dtype: int64

For the cumulative production values, I decided to drop all NaN's, as these wells probably don't have data for 365 days, or don't have data. In either case, they can't go into the machine learning model:

In [9]:
prods_365 = prods_365.dropna()

In [10]:
prods_365.index.nunique() # number of cumulative 365 production data points

6459

# Completions Data

The completions data had several duplicate rows, and also wells which appeared to have been fracked twice.

1) I first eliminated the duplicate rows (rows which have all values duplicated, then rows which appeared to have been fracked twice in the same day). 

2) Then I counted how many wells had been refracked, and dropped them as they were not too many. Figuring out how to handle refracked wells would be an interesting problem, but there weren't enough data points to justify the exercise.

In [11]:
# check original number of data points in completions
completions.shape

(9664, 16)

In [12]:
# eliminate rows in completions data where all column values are the same
completions.drop_duplicates(inplace=True)
completions.shape

(9663, 16)

In [13]:
# eliminate rows where both api and treatment date are the same
completions.drop_duplicates(subset=['api','treatment_date'],keep=False,inplace=True)
completions.shape

(9575, 16)

In [14]:
# list rows where the api's are the same, but the treatment dates are different
# these are wells which have been refracked
completions[completions.duplicated(subset=['api'],keep=False)]

api  bottom  file_number  fluid_bbl  fluid_bbl_per_ft  \
11    33025903100000    5800        90310         47                 0   
13    33025903100000    5690        90310         95                 1   
99    33061031180000   14795        28450      21588                 7   
189   33025010440000   20725        18681      41412                 4   
205   33025008660000   19414        17804      42265                 4   
216   33089006800000    8054        22223         35                 4   
218   33023010900000   18011        26095      40379                 4   
231   33023011110000   14385        26583      28991                 4   
239   33061005630000   19711        16772      20450                 2   
295   33053054560000   19445        26927      52920                 6   
476   33053042370000   20763        23301      76061                 8   
493   33089006800000    8054        22223        547                68   
500   33061028740000   17816        27224      77367                11   
650   33075014620000    4596        28329          2                 0   
685   33025008270000   19605        17599      50586                 5   
687   33061031180000   15550        28450      27591                 6   
753   33075014620000    4603        28329          5                 1   
858   33025008510000   20073        17753      39398                 4   
967   33061028740000   21095        27224      27694                 8   
1101  33007018180000   20568        27543      83535                 9   
1129  33061009820000   20098        17966      38346                 4   
1181  33061009090000   19918        17771      42260                 4   
1286  33061019350000   19320        21957      75535                 8   
1317  33053043630000   13788        23673      11283                 5   
1332  33053043140000   16822        23544      59367                10   
1335  33009023540000    4479        27319         54                 0   
1336  33089006800000    8054        22223        166                20   
1358  33009023540000    4479        27319         23                 2   
1364  33053051780000   21385        26136      42879                 4   
1420  33053051060000   14436        25892      40657                10   
...              ...     ...          ...        ...               ...   
9448  33061009820000   20098        17966      18645                 2   
9462  33023005230000   13420        16794       1524                 0   
9463  33023005230000   14059        16794       1524                 3   
9466  33009021610000    3392        17878         35                 1   
9467  33009021610000    3392        17878         11                 0   
9468  33061005630000   19711        16772      20451                 2   
9478  33053015420000    9452         9618        452                 2   
9479  33053030060000   14693        17974       1858                 2   
9480  33053030060000   19330        17974       8774                 2   
9490  33025008780000   20095        17865      10763                 1   
9493  33053030010000   19456        17952      14238                 1   
9502  33053022930000   13841        12699         28                 0   
9504  33007013740000    8131        14277         23                 1   
9521  33033002890000   12031        17519       3829                 4   
9524  33025008450000   20080        17713      13448                 1   
9525  33025008440000   20141        17712      19702                 2   
9556  33053015420000    9452         9618         85                 0   
9560  33061008920000   19874        17709      19422                 2   
9576  33025008470000   18925        17721      12697                 1   
9587  33053015960000    5910         9774         48                 0   
9609  33061009090000   19906        17771      19920                 2   
9615  33105016910000   13520        17488         35                 2   
9617  33105016910000 

In [15]:
# this gives the total number of refracked wells
completions[completions.duplicated(subset=['api'],keep=False)]['api'].nunique()

348

In [16]:
# dropping refracked wells
completions.drop_duplicates(subset=['api'],keep=False,inplace=True)
completions.shape

(8662, 16)

Therefore, the number of data points lost from completions is 9664-8662 = 1002

# Prepare final data frame for machine learning models

The production, completions, and well_indices dataframes are merged here to run machine learning models.

In [17]:
prods_365.reset_index(inplace=True)

In [18]:
final_data = pd.merge(prods_365, well_indices,how='inner',on='api')

In [19]:
final_data = pd.merge(final_data, completions,how='inner',on='api')

In [20]:
final_data.head()

api  cum_oil_365  file_no                    Current_Operator  \
0  33053032160000     262305.0    19323                  QEP ENERGY COMPANY   
1  33053036170000     151827.0    20892                 EOG RESOURCES, INC.   
2  33025010660000     168315.0    18809  ENERPLUS RESOURCES USA CORPORATION   
3  33053051550000      84342.0    26055     HESS BAKKEN INVESTMENTS II, LLC   
4  33053043160000     104623.0    23550                  HRC OPERATING, LLC   

                 Current_Well_Name     Lease_Name      Lease_Number  \
0               THOMPSON  1-29/32H       THOMPSON          1-29/32H   
1           CLARKS CREEK  16-0706H   CLARKS CREEK          16-0706H   
2                    BAKER  20-34H          BAKER            20-34H   
3        BW-THELMA  150-99-3031H-3      BW-THELMA    150-99-3031H-3   
4  FORT BERTHOLD  152-94-14C-11-2H  FORT BERTHOLD  152-94-14C-11-2H   

                 Original_Operator               Original_Well_Name  \
0  HELIS OIL & GAS COMPANY, L.L.C.               THOMPSON  1-29/32H   
1              EOG RESOURCES, INC.           CLARKS CREEK  16-0706H   
2           PEAK NORTH DAKOTA, LLC                    BAKER  20-34H   
3                 HESS CORPORATION        BW-THELMA  150-99-3031H-3   
4               PETRO-HUNT, L.L.C.  FORT BERTHOLD  152-94-14C-11-2H   

    Spud_Date      ...        fluid_gal_per_ft ft_per_stage lateral_length  \
0  01/25/2011      ...                     366         9400           9400   
1  09/06/2011      ...                     385          221           6875   
2  10/07/2010      ...                     109         8597           8597   
3  09/07/2013      ...                     175          323           9692   
4  10/27/2012      ...                     328          351           9133   

  max_treat_press  max_treat_rate propp_lbs propp_lbs_per_ft stages    top  \
0            8338            38.5   3285650              349      1  11148   
1            8762            32.3    363915               52     31  11205   
2            9372            40.9   2807153              326      1  11080   
3            8206            25.6   1557517              160     30  11433   
4            8673            39.1   2625720              287     26  11386   

  treatment_date  
0        4/26/11  
1        1/14/12  
2         8/1/11  
3       12/17/13  
4       11/28/12  

[5 rows x 41 columns]

In [21]:
final_data.shape

(5729, 41)

In [22]:
final_data.columns

Index(['api', 'cum_oil_365', 'file_no', 'Current_Operator',
       'Current_Well_Name', 'Lease_Name', 'Lease_Number', 'Original_Operator',
       'Original_Well_Name', 'Spud_Date', 'TD', 'County_Name', 'Township',
       'Range', 'Section', 'QQ', 'Footages', 'Field_Name', 'Produced_Pools',
       'Wellbore', 'Latitude', 'Longitude', 'Well_Type', 'Well_Status', 'CTB',
       'Well_Status_Date', 'bottom', 'file_number', 'fluid_bbl',
       'fluid_bbl_per_ft', 'fluid_gal', 'fluid_gal_per_ft', 'ft_per_stage',
       'lateral_length', 'max_treat_press', 'max_treat_rate', 'propp_lbs',
       'propp_lbs_per_ft', 'stages', 'top', 'treatment_date'],
      dtype='object')

In [23]:
# checking whether the fluid gallons and barrels are indeed equivalent
# Use 1 barrel = 42 fluid gallon
fluid_gal = final_data.fluid_gal
fluid_bbl = final_data.fluid_bbl

gal_convert = fluid_gal/42

# getting the max error in conversion
abs(fluid_bbl - gal_convert).max()


0.97619047619082266

This is a pretty low value considering I'm working with very large amounts of fluid

In [24]:
# Similarly, checking if the difference between the top and bottom columns corresponds to lateral length
tops = final_data.top
bots = final_data.bottom
lats_calc = bots - tops
lats = final_data.lateral_length
abs(lats - lats_calc).max()


0

## Columns/rows to drop from final data frame

We can now decide which columns are redundant in the final data frame, and should be removed

1) The fluid in gallons and barrels are both reported, so obviously one is redundant. We keep the fluid measured in barrels, and lose the ['fluid_gal','fluid_gal_per_ft'] columns

2) There are file numbers, which should have no bearing on the machine learning model (columns 'file_no')

3) The Current_Well_Name and Lease_Number do not seem to have value for a machine learning model.

4) There are a number of columns related to the spatial location of the well. While spatial location should be very important, these columns denote geographical rather than geological information. So, I dropped ['Township','Range','Section']

5) Other names are related to how the oil field was named. This might be important, but I have decided to keep all that information within latitude and longitude. So I dropped ['Lease_Name','Field_Name','Original_Well_Name','Original_Operator','Current_Operator']

6) I am not sure of the meaning of these columns, and have decided to drop them as they are categorical variables which could have some meaning of no importance to the machine learning model: ['Well_Status','Well_Status_Date','QQ','CTB']

7) The difference between top and bottom gives the lateral length. We do not need all three variables. I decided to drop top.


In [25]:
cols_to_drop = ['file_no','Current_Well_Name','Lease_Number','Township','Well_Type','Well_Status_Date','file_number','fluid_gal','fluid_gal_per_ft','top','Footages','Lease_Name','Field_Name','QQ','Range','CTB','Section','Well_Status','Original_Well_Name','Original_Operator','Current_Operator']
final_data.drop(labels=cols_to_drop,axis=1,inplace=True)

In [26]:
final_data.columns

Index(['api', 'cum_oil_365', 'Spud_Date', 'TD', 'County_Name',
       'Produced_Pools', 'Wellbore', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'treatment_date'],
      dtype='object')

In [27]:
final_data.head()

api  cum_oil_365   Spud_Date       TD County_Name  \
0  33053032160000     262305.0  01/25/2011  20548.0    MCKENZIE   
1  33053036170000     151827.0  09/06/2011  18748.0    MCKENZIE   
2  33025010660000     168315.0  10/07/2010  19990.0        DUNN   
3  33053051550000      84342.0  09/07/2013  21125.0    MCKENZIE   
4  33053043160000     104623.0  10/27/2012  20790.0    MCKENZIE   

  Produced_Pools    Wellbore   Latitude   Longitude  bottom  fluid_bbl  \
0         BAKKEN  HORIZONTAL  47.788616 -102.860602   20548      81925   
1         BAKKEN  HORIZONTAL  47.905914 -102.751654   18080      63071   
2         BAKKEN  HORIZONTAL  47.703757 -102.481510   19677      22445   
3         BAKKEN  HORIZONTAL  47.788427 -103.398873   21125      40552   
4         SANISH  HORIZONTAL  47.978685 -102.681657   20519      71516   

   fluid_bbl_per_ft  ft_per_stage  lateral_length  max_treat_press  \
0                 8          9400            9400             8338   
1                 9           221            6875             8762   
2                 2          8597            8597             9372   
3                 4           323            9692             8206   
4                 7           351            9133             8673   

   max_treat_rate  propp_lbs  propp_lbs_per_ft  stages treatment_date  
0            38.5    3285650               349       1        4/26/11  
1            32.3     363915                52      31        1/14/12  
2            40.9    2807153               326       1         8/1/11  
3            25.6    1557517               160      30       12/17/13  
4            39.1    2625720               287      26       11/28/12

In [28]:
# get number of null entries in final_data
final_data.isnull().sum()

api                 0
cum_oil_365         0
Spud_Date           0
TD                  7
County_Name         0
Produced_Pools      0
Wellbore            5
Latitude            0
Longitude           0
bottom              0
fluid_bbl           0
fluid_bbl_per_ft    0
ft_per_stage        0
lateral_length      0
max_treat_press     0
max_treat_rate      0
propp_lbs           0
propp_lbs_per_ft    0
stages              0
treatment_date      0
dtype: int64

Since there ain't that many null values in any column, I drop the rows with null values

In [29]:
final_data.dropna(inplace=True)

## Handling categorical variables and dates

We now have to deal with the categorical variables and dates. 

The dates are converted to number of days from a reference date. I picked today's date, May 23rd, 2018, and converted both spud date and treatment_date to these time delta's.

In [30]:
today_date = datetime.date(2018, 5, 23)
Spud_delta = (today_date - pd.to_datetime(final_data.Spud_Date)).dt.days
final_data['spud_timedelta'] = Spud_delta

In [31]:
treatment_delta = (today_date - pd.to_datetime(final_data.treatment_date)).dt.days
final_data['treatment_timedelta'] = treatment_delta

In [32]:
final_data.head()

api  cum_oil_365   Spud_Date       TD County_Name  \
0  33053032160000     262305.0  01/25/2011  20548.0    MCKENZIE   
1  33053036170000     151827.0  09/06/2011  18748.0    MCKENZIE   
2  33025010660000     168315.0  10/07/2010  19990.0        DUNN   
3  33053051550000      84342.0  09/07/2013  21125.0    MCKENZIE   
4  33053043160000     104623.0  10/27/2012  20790.0    MCKENZIE   

  Produced_Pools    Wellbore   Latitude   Longitude  bottom  \
0         BAKKEN  HORIZONTAL  47.788616 -102.860602   20548   
1         BAKKEN  HORIZONTAL  47.905914 -102.751654   18080   
2         BAKKEN  HORIZONTAL  47.703757 -102.481510   19677   
3         BAKKEN  HORIZONTAL  47.788427 -103.398873   21125   
4         SANISH  HORIZONTAL  47.978685 -102.681657   20519   

          ...          ft_per_stage  lateral_length  max_treat_press  \
0         ...                  9400            9400             8338   
1         ...                   221            6875             8762   
2         ...                  8597            8597             9372   
3         ...                   323            9692             8206   
4         ...                   351            9133             8673   

   max_treat_rate  propp_lbs  propp_lbs_per_ft  stages  treatment_date  \
0            38.5    3285650               349       1         4/26/11   
1            32.3     363915                52      31         1/14/12   
2            40.9    2807153               326       1          8/1/11   
3            25.6    1557517               160      30        12/17/13   
4            39.1    2625720               287      26        11/28/12   

   spud_timedelta treatment_timedelta  
0            2675                2584  
1            2451                2321  
2            2785                2487  
3            1719                1618  
4            2034                2002  

[5 rows x 22 columns]

We can now safely drop the Spud_date and treatment_date from the dateframe

In [33]:
final_data.drop(['Spud_Date','treatment_date'],axis=1,inplace=True)

In [34]:
final_data.columns

Index(['api', 'cum_oil_365', 'TD', 'County_Name', 'Produced_Pools', 'Wellbore',
       'Latitude', 'Longitude', 'bottom', 'fluid_bbl', 'fluid_bbl_per_ft',
       'ft_per_stage', 'lateral_length', 'max_treat_press', 'max_treat_rate',
       'propp_lbs', 'propp_lbs_per_ft', 'stages', 'spud_timedelta',
       'treatment_timedelta'],
      dtype='object')

The categorical variables are turned into a format suitable for ML models by using the one hot encoding feature in pandas.

In [35]:
final_data = pd.get_dummies(final_data)


In [36]:
final_data.columns

Index(['api', 'cum_oil_365', 'TD', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'spud_timedelta', 'treatment_timedelta',
       'County_Name_BILLINGS', 'County_Name_BOTTINEAU', 'County_Name_BURKE',
       'County_Name_DIVIDE', 'County_Name_DUNN', 'County_Name_GOLDEN VALLEY',
       'County_Name_MCKENZIE', 'County_Name_MCLEAN', 'County_Name_MERCER',
       'County_Name_MOUNTRAIL', 'County_Name_RENVILLE', 'County_Name_STARK',
       'County_Name_WILLIAMS', 'Produced_Pools_BAKKEN',
       'Produced_Pools_BAKKEN/THREE FORKS', 'Produced_Pools_BIRDBEAR',
       'Produced_Pools_LODGEPOLE', 'Produced_Pools_MADISON',
       'Produced_Pools_ORDOVICIAN', 'Produced_Pools_RED RIVER',
       'Produced_Pools_SANISH', 'Produced_Pools_SPEARFISH',
       'Produced_Pools_SPEARFISH/MADISON', 'Produced_Pools_THREE FORKS',
       'Produced_Pools_TYLER'

In [37]:
final_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5717 entries, 0 to 5728
Data columns (total 46 columns):
api                                  5717 non-null int64
cum_oil_365                          5717 non-null float64
TD                                   5717 non-null float64
Latitude                             5717 non-null float64
Longitude                            5717 non-null float64
bottom                               5717 non-null int64
fluid_bbl                            5717 non-null int64
fluid_bbl_per_ft                     5717 non-null int64
ft_per_stage                         5717 non-null int64
lateral_length                       5717 non-null int64
max_treat_press                      5717 non-null int64
max_treat_rate                       5717 non-null float64
propp_lbs                            5717 non-null int64
propp_lbs_per_ft                     5717 non-null int64
stages                               5717 non-null int64
spud_timedelta              

We see all the columns have the same number of non-null entries, which should mean no unexpected null values.

In [38]:
non_cat_cols = ['api', 'cum_oil_365', 'TD', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'spud_timedelta', 'treatment_timedelta']

## Data cleaning/sanity check

In [39]:
final_data.describe(exclude=['uint8']) # excluding uint8 values removes the categorical variables from view

api    cum_oil_365            TD     Latitude    Longitude  \
count  5.717000e+03    5717.000000   5717.000000  5717.000000  5717.000000   
mean   3.305595e+13   90386.078538  19323.071016    48.019745  -102.949297   
std    2.889531e+10   49320.821437   2674.805085     0.469631     0.554453   
min    3.300702e+13     773.000000   1860.000000    46.736696  -104.044496   
25%    3.302502e+13   57538.000000  18972.000000    47.750768  -103.362034   
50%    3.305305e+13   82617.000000  20210.000000    48.024951  -102.911239   
75%    3.306102e+13  113261.000000  20774.000000    48.285520  -102.549788   
max    3.310503e+13  501641.000000  26908.000000    48.997699  -100.553543   

              bottom     fluid_bbl  fluid_bbl_per_ft   ft_per_stage  \
count    5717.000000  5.717000e+03       5717.000000    5717.000000   
mean    19310.900822  5.223893e+04         16.333567    3358.932482   
std      5520.828737  4.193443e+04        163.076524    4416.695004   
min      2970.000000  0.000000e+00          0.000000       0.000000   
25%     18825.000000  3.017200e+04          3.000000     300.000000   
50%     20118.000000  4.767500e+04          5.000000     348.000000   
75%     20709.000000  6.451300e+04          7.000000    8849.000000   
max    232035.000000  1.514920e+06       5265.000000  131211.000000   

       lateral_length  max_treat_press  max_treat_rate     propp_lbs  \
count     5717.000000      5717.000000     5717.000000  5.717000e+03   
mean      8697.700367      7563.977436       54.839706  2.848323e+06   
std       5246.815869      3302.874885      403.664456  4.509289e+06   
min          2.000000         0.000000        0.000000  0.000000e+00   
25%       8696.000000      7230.000000       29.800000  1.864033e+06   
50%       9342.000000      8252.000000       35.600000  2.611720e+06   
75%       9688.000000      8839.000000       41.200000  3.381308e+06   
max     220106.000000    101956.000000     9403.000000  2.878837e+08   

       propp_lbs_per_ft       stages  spud_timedelta  treatment_timedelta  
count       5717.000000  5717.000000     5717.000000          5717.000000  
mean         958.599790    18.273920     2360.998601          2250.151128  
std         9497.750949    14.785515      422.196264           427.166841  
min            0.000000     1.000000     1575.000000          1286.000000  
25%          222.000000     1.000000     2014.000000          1892.000000  
50%          298.000000    25.000000     2295.000000          2190.000000  
75%          391.000000    30.000000     2675.000000          2533.000000  
max       304021.000000    71.000000     3398.000000          3387.000000

From the max and min values for some columns, it is clear some values are not right. For example, a lateral length of 220106 feet seems too much.

In [40]:
final_data[final_data['lateral_length']>20000]['lateral_length']

323     131211
2031    174558
4146     86601
4751    220106
5389    199728
Name: lateral_length, dtype: int64

Just 5 wells seem to be of extraordinary lateral length. Dropping these 5:

In [41]:
final_data = final_data[final_data['lateral_length']<20000]

In [42]:
final_data.describe(exclude=['uint8'])

api    cum_oil_365            TD     Latitude    Longitude  \
count  5.712000e+03    5712.000000   5712.000000  5712.000000  5712.000000   
mean   3.305595e+13   90385.081933  19323.295693    48.019605  -102.949278   
std    2.889438e+10   49330.072131   2674.715095     0.469580     0.554603   
min    3.300702e+13     773.000000   1860.000000    46.736696  -104.044496   
25%    3.302502e+13   57531.250000  18972.000000    47.750749  -103.362221   
50%    3.305305e+13   82605.500000  20210.000000    48.024937  -102.911222   
75%    3.306102e+13  113278.000000  20774.000000    48.285508  -102.549727   
max    3.310503e+13  501641.000000  26908.000000    48.997699  -100.553543   

             bottom     fluid_bbl  fluid_bbl_per_ft  ft_per_stage  \
count   5712.000000  5.712000e+03       5712.000000   5712.000000   
mean   19176.473915  5.224543e+04         16.347864   3335.203782   
std     2783.331166  4.195020e+04        163.147179   4081.301392   
min     2970.000000  0.000000e+00          0.000000      0.000000   
25%    18823.250000  3.017000e+04          3.000000    300.000000   
50%    20117.500000  4.767750e+04          5.000000    348.000000   
75%    20704.250000  6.456800e+04          7.000000   8849.500000   
max    26908.000000  1.514920e+06       5265.000000  14576.000000   

       lateral_length  max_treat_press  max_treat_rate     propp_lbs  \
count     5712.000000      5712.000000     5712.000000  5.712000e+03   
mean      8563.121324      7565.665091       54.865984  2.848648e+06   
std       2193.519433      3302.580119      403.839975  4.511218e+06   
min          2.000000         0.000000        0.000000  0.000000e+00   
25%       8696.000000      7232.000000       29.875000  1.864024e+06   
50%       9341.000000      8253.000000       35.600000  2.611664e+06   
75%       9686.250000      8839.250000       41.300000  3.382742e+06   
max      19858.000000    101956.000000     9403.000000  2.878837e+08   

       propp_lbs_per_ft       stages  spud_timedelta  treatment_timedelta  
count       5712.000000  5712.000000     5712.000000          5712.000000  
mean         959.423669    18.267157     2361.037290          2250.211485  
std         9501.866844    14.785446      422.214931           427.159516  
min            0.000000     1.000000     1575.000000          1286.000000  
25%          222.000000     1.000000     2014.000000          1892.000000  
50%          299.000000    25.000000     2295.000000          2190.500000  
75%          391.000000    30.000000     2675.000000          2533.000000  
max       304021.000000    71.000000     3398.000000          3387.000000

Some lateral length values are also really small, resulting in very high 'per_ft' values for some variables. It is possible these are associated with vertical wells. 

In [43]:
final_data[(final_data['lateral_length']<5)][['cum_oil_365', 'TD', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'spud_timedelta', 'treatment_timedelta','Wellbore_DIRECTIONAL', 'Wellbore_HORIZONTAL',
       'Wellbore_HORIZONTAL RE-ENTRY', 'Wellbore_VERTICAL']].describe()

cum_oil_365            TD   Latitude   Longitude       bottom  \
count      6.000000      6.000000   6.000000    6.000000      6.00000   
mean    9359.500000   8751.166667  48.252224 -102.151373   8467.50000   
std     5058.695988   6582.298456   0.871788    1.182753   6250.27461   
min     3390.000000   3204.000000  46.736696 -103.553665   3072.00000   
25%     5990.750000   4835.500000  47.923657 -103.110601   4722.50000   
50%     8269.000000   6460.000000  48.582655 -102.054398   6336.00000   
75%    13154.250000   9877.000000  48.838859 -101.394671   9686.50000   
max    16215.000000  21073.000000  48.983086 -100.608344  20073.00000   

         fluid_bbl  fluid_bbl_per_ft  ft_per_stage  lateral_length  \
count     6.000000          6.000000      6.000000        6.000000   
mean    737.000000        245.000000      3.333333        3.333333   
std    1237.280405        412.893207      0.816497        0.816497   
min       2.000000          0.000000      2.000000        2.000000   
25%       6.000000          2.000000      3.000000        3.000000   
50%      16.000000          3.500000      3.500000        3.500000   
75%    1043.000000        347.000000      4.000000        4.000000   
max    3000.000000       1000.000000      4.000000        4.000000   

       max_treat_press  max_treat_rate      propp_lbs  propp_lbs_per_ft  \
count         6.000000        6.000000       6.000000          6.000000   
mean       2131.833333       11.666667   35633.333333      11877.666667   
std        3193.328196       17.844626   66349.518963      22116.363107   
min           0.000000        0.000000       0.000000          0.000000   
25%         150.500000        0.025000       0.000000          0.000000   
50%         500.000000        0.800000       0.000000          0.000000   
75%        2975.000000       21.675000   36525.000000      12174.750000   
max        7957.000000       40.000000  165100.000000      55033.000000   

       stages  spud_timedelta  treatment_timedelta  Wellbore_DIRECTIONAL  \
count     6.0        6.000000             6.000000                   6.0   
mean      1.0     2117.000000          2046.000000                   0.0   
std       0.0      260.034613           243.599672                   0.0   
min       1.0     1889.000000          1832.000000                   0.0   
25%       1.0     1933.250000          1889.500000                   0.0   
50%       1.0     2007.000000          1959.000000                   0.0   
75%       1.0     2300.500000          2126.000000                   0.0   
max       1.0     2492.000000          2478.000000                   0.0   

       Wellbore_HORIZONTAL  Wellbore_HORIZONTAL RE-ENTRY  Wellbore_VERTICAL  
count             6.000000                           6.0           6.000000  
mean              0.333333                           0.0           0.666667  
std               0.516398                           0.0           0.516398  
min               0.000000                           0.0           0.000000  
25%               0.000000                           0.0           0.250000  
50%               0.000000                           0.0           1.000000  
75%               0.750000                           0.0           1.000000  
max               1.000000                           0.0           1.000000

5 feet seems like a really small lateral length for a horizontal well, so dropping these rows

In [44]:
final_data = final_data[(final_data['lateral_length']>5)] #| (final_data['Wellbore_HORIZONTAL']==0)]

Some of the maximum treatment pressure values are really high. 101956 psi should be too high a value.

In [45]:
final_data[final_data.max_treat_press > 20000][non_cat_cols]

api  cum_oil_365       TD   Latitude   Longitude  bottom  \
340   33089006440000     163064.0  20394.0  46.947868 -103.208196   20394   
646   33025019780000      46846.0  18787.0  47.506111 -102.346901   18787   
1291  33105025290000     112777.0  21050.0  48.082695 -103.477887   21050   
3459  33105030270000     102155.0  20816.0  48.343517 -103.570917   20754   
5489  33007016560000      36053.0  19365.0  47.078151 -103.428707   19365   

      fluid_bbl  fluid_bbl_per_ft  ft_per_stage  lateral_length  \
340       28870                 3          9271            9271   
646       32088                 4           257            7733   
1291      89866                 9           268            9680   
3459      92369                 9           320            9945   
5489      32929                 3          8541            8541   

      max_treat_press  max_treat_rate  propp_lbs  propp_lbs_per_ft  stages  \
340             58369             0.0    1747380               188       1   
646             77961            37.7    2011540               260      30   
1291           101956            36.6    4457382               460      36   
3459            76946            31.3    4057871               408      31   
5489            94554             0.0    1837840               215       1   

      spud_timedelta  treatment_timedelta  
340             2430                 2361  
646             2004                 1965  
1291            2196                 2108  
3459            1844                 1698  
5489            2388                 2314

In [46]:
final_data = final_data[final_data.max_treat_press < 20000] # eliminating very high max treat pressure values

In [47]:
final_data.describe(exclude=['uint8'])

api    cum_oil_365            TD     Latitude    Longitude  \
count  5.700000e+03    5700.000000   5700.000000  5700.000000  5700.000000   
mean   3.305599e+13   90482.260526  19336.293333    48.019637  -102.950166   
std    2.885699e+10   49281.976835   2641.118919     0.468794     0.552877   
min    3.300702e+13     773.000000   1860.000000    46.804382  -104.044496   
25%    3.302502e+13   57670.000000  18981.000000    47.755395  -103.362034   
50%    3.305305e+13   82696.500000  20210.000000    48.024921  -102.911222   
75%    3.306102e+13  113336.000000  20774.000000    48.285486  -102.550259   
max    3.310503e+13  501641.000000  26908.000000    48.997699  -100.553543   

             bottom     fluid_bbl  fluid_bbl_per_ft  ft_per_stage  \
count   5700.000000  5.700000e+03       5700.000000   5700.000000   
mean   19189.498421  5.230620e+04         16.119298   3338.947719   
std     2751.374150  4.194621e+04        162.690601   4081.984876   
min     2970.000000  0.000000e+00          0.000000      0.000000   
25%    18827.500000  3.022000e+04          3.000000    300.000000   
50%    20118.500000  4.775500e+04          5.000000    348.000000   
75%    20705.000000  6.459900e+04          7.000000   8857.000000   
max    26908.000000  1.514920e+06       5265.000000  14576.000000   

       lateral_length  max_treat_press  max_treat_rate     propp_lbs  \
count     5700.000000      5700.000000     5700.000000  5.700000e+03   
mean      8573.220000      7507.421404       54.950649  2.852132e+06   
std       2175.040798      2412.232877      404.259924  4.514748e+06   
min          6.000000         0.000000        0.000000  0.000000e+00   
25%       8696.750000      7233.750000       29.900000  1.866527e+06   
50%       9342.000000      8253.000000       35.650000  2.613570e+06   
75%       9687.000000      8838.000000       41.300000  3.383022e+06   
max      19858.000000     19788.000000     9403.000000  2.878837e+08   

       propp_lbs_per_ft       stages  spud_timedelta  treatment_timedelta  
count       5700.000000  5700.000000     5700.000000          5700.000000  
mean         948.672105    18.287018     2361.397193          2250.498246  
std         9482.630219    14.781431      422.399580           427.377728  
min            0.000000     1.000000     1575.000000          1286.000000  
25%          222.000000     1.000000     2015.000000          1892.000000  
50%          299.000000    25.000000     2295.500000          2191.000000  
75%          391.000000    30.000000     2675.000000          2533.000000  
max       304021.000000    71.000000     3398.000000          3387.000000

In [48]:
(final_data.propp_lbs_per_ft > 50000).sum()

24

In [61]:
final_data = final_data[final_data.propp_lbs_per_ft < 50000]

Some max_treat_rate values are really high:

In [62]:
final_data[final_data.max_treat_rate > 5000][non_cat_cols]

api  cum_oil_365       TD   Latitude   Longitude  bottom  \
249   33053043880000      27501.0  20954.0  47.760695 -103.365855   20954   
300   33089006830000      53554.0  20203.0  46.918925 -103.095995   20203   
396   33089006560000      76215.0  21418.0  46.949866 -103.133397   21418   
450   33053032750000      70931.0  19852.0  47.791476 -103.363595   19852   
698   33025009710000     149470.0  19930.0  47.747671 -102.505411   19930   
819   33105020540000      65336.0  20436.0  48.285291 -103.813531   20436   
2644  33053047740000      86506.0  21388.0  47.818074 -103.179295   21388   
2708  33025016550000      85879.0  20318.0  47.182298 -102.886126   20059   
2937  33053042090000      62178.0  20610.0  47.878315 -103.681742   20610   
3395  33061021640000     102232.0  19824.0  48.293684 -102.549116   19824   
4171  33025015840000      47615.0  20740.0  47.097486 -102.765545   20487   
4291  33025015980000     151831.0  20430.0  47.268909 -102.902880   20227   
4325  33105031290000      70817.0  19995.0  48.458892 -103.822558   19784   
4419  33061021760000     116731.0  17278.0  48.131060 -102.481567   17278   

      fluid_bbl  fluid_bbl_per_ft  ft_per_stage  lateral_length  \
249       47866                 5           337            9454   
300       43597                 4           314            9447   
396       39590                 3           347           10431   
450       55512                 6           332            8306   
698       52350                 5          8841            8841   
819       74813                 7          9593            9593   
2644      81747                 8           334           10041   
2708      48610                 5           420            8822   
2937      84093                 8           258            9569   
3395      70099                 7           316            9504   
4171      67084                 7           451            9484   
4291      68950                 7           436            9162   
4325     192937                20           268            9394   
4419      24526                 3           313            6904   

      max_treat_press  max_treat_rate  propp_lbs  propp_lbs_per_ft  stages  \
249              9200          7369.0    2347343               248      28   
300              9641          5418.0    3001390               317      30   
396              8344          6970.0    2356214               225      30   
450              7052          7052.0    2596600               312      25   
698              9029          6594.0    3340438               377       1   
819              8998          6943.0    4391594               457       1   
2644               37          8690.0    3436308               342      30   
2708               54          8663.0    1665744               188      21   
2937               40          8062.0    3145840               328      37   
3395               40          7797.0    3700297               389      30   
4171               52          8015.0    1474916               155      21   
4291               61          8158.0    1704925               186      21   
4325               77          9403.0    3982280               423      35   
4419             8812          5856.0    1713499               248      22   

      spud_timedelta  treatment_timedelta  
249             1959                 1835  
300             2192                 2136  
396             2224                 2138  
450             2669                 2345  
698             3079                 2917  
819             2672                 2494  
2644            1901                 1728  
2708            2223                 2010  
2937            2128                 2058  
3395            1904                 1706  
4171            2263                 2130  
4291            2277                 2184  
4325            1741                 1672  
4419            2089                 2042

Some of the correspondinh max_treat_pressure values have very low values associated with them. I will remove the unreasonable pairings

In [63]:
final_data= final_data[(final_data['max_treat_rate'] < 7500) | (final_data['max_treat_press'] < 100)] # this value corresponds to the unreasonable pairings

In [64]:
final_data.describe(exclude=['uint8'])

api    cum_oil_365            TD     Latitude    Longitude  \
count  5.582000e+03    5582.000000   5582.000000  5582.000000  5582.000000   
mean   3.305615e+13   91134.686313  19433.648692    48.013984  -102.960521   
std    2.858789e+10   49186.193236   2447.968144     0.462773     0.527609   
min    3.300702e+13     773.000000   1860.000000    46.804382  -104.044496   
25%    3.302502e+13   58313.000000  19039.250000    47.750153  -103.364224   
50%    3.305305e+13   83211.000000  20228.500000    48.023947  -102.909950   
75%    3.306102e+13  113831.500000  20780.000000    48.283580  -102.550748   
max    3.310503e+13  501641.000000  26908.000000    48.997699  -100.553543   

             bottom     fluid_bbl  fluid_bbl_per_ft  ft_per_stage  \
count   5582.000000  5.582000e+03       5582.000000   5582.000000   
mean   19331.061985  5.316869e+04          6.764780   3363.822465   
std     2481.937223  4.183591e+04         29.743881   4092.174083   
min     2970.000000  5.000000e+00          0.000000      2.000000   
25%    18912.250000  3.082700e+04          3.000000    302.000000   
50%    20139.500000  4.822600e+04          5.000000    350.000000   
75%    20716.000000  6.524700e+04          7.000000   8882.000000   
max    26908.000000  1.514920e+06        875.000000  14576.000000   

       lateral_length  max_treat_press  max_treat_rate     propp_lbs  \
count     5582.000000      5582.000000     5582.000000  5.582000e+03   
mean      8672.439627      7570.493013       52.964529  2.886566e+06   
std       2017.239020      2351.649084      381.769868  4.551447e+06   
min          6.000000         0.000000        0.000000  0.000000e+00   
25%       8752.250000      7313.500000       30.000000  1.883984e+06   
50%       9353.000000      8275.000000       35.850000  2.625670e+06   
75%       9695.750000      8845.750000       41.300000  3.405264e+06   
max      19858.000000     19788.000000     9403.000000  2.878837e+08   

       propp_lbs_per_ft       stages  spud_timedelta  treatment_timedelta  
count       5582.000000  5582.000000     5582.000000          5582.000000  
mean         396.179326    18.355428     2362.873522          2251.469724  
std         1708.620071    14.797316      422.057178           427.529246  
min            0.000000     1.000000     1575.000000          1286.000000  
25%          225.000000     1.000000     2017.000000          1894.000000  
50%          300.000000    25.000000     2298.000000          2192.000000  
75%          391.000000    30.000000     2676.000000          2533.000000  
max        49725.000000    71.000000     3398.000000          3387.000000

It also doesn't make sense for there to be no fluid at all for a frack job.

In [65]:
final_data = final_data[final_data['fluid_bbl'] > 0]

In [66]:
final_data.describe(exclude=['uint8'])

api    cum_oil_365            TD     Latitude    Longitude  \
count  5.582000e+03    5582.000000   5582.000000  5582.000000  5582.000000   
mean   3.305615e+13   91134.686313  19433.648692    48.013984  -102.960521   
std    2.858789e+10   49186.193236   2447.968144     0.462773     0.527609   
min    3.300702e+13     773.000000   1860.000000    46.804382  -104.044496   
25%    3.302502e+13   58313.000000  19039.250000    47.750153  -103.364224   
50%    3.305305e+13   83211.000000  20228.500000    48.023947  -102.909950   
75%    3.306102e+13  113831.500000  20780.000000    48.283580  -102.550748   
max    3.310503e+13  501641.000000  26908.000000    48.997699  -100.553543   

             bottom     fluid_bbl  fluid_bbl_per_ft  ft_per_stage  \
count   5582.000000  5.582000e+03       5582.000000   5582.000000   
mean   19331.061985  5.316869e+04          6.764780   3363.822465   
std     2481.937223  4.183591e+04         29.743881   4092.174083   
min     2970.000000  5.000000e+00          0.000000      2.000000   
25%    18912.250000  3.082700e+04          3.000000    302.000000   
50%    20139.500000  4.822600e+04          5.000000    350.000000   
75%    20716.000000  6.524700e+04          7.000000   8882.000000   
max    26908.000000  1.514920e+06        875.000000  14576.000000   

       lateral_length  max_treat_press  max_treat_rate     propp_lbs  \
count     5582.000000      5582.000000     5582.000000  5.582000e+03   
mean      8672.439627      7570.493013       52.964529  2.886566e+06   
std       2017.239020      2351.649084      381.769868  4.551447e+06   
min          6.000000         0.000000        0.000000  0.000000e+00   
25%       8752.250000      7313.500000       30.000000  1.883984e+06   
50%       9353.000000      8275.000000       35.850000  2.625670e+06   
75%       9695.750000      8845.750000       41.300000  3.405264e+06   
max      19858.000000     19788.000000     9403.000000  2.878837e+08   

       propp_lbs_per_ft       stages  spud_timedelta  treatment_timedelta  
count       5582.000000  5582.000000     5582.000000          5582.000000  
mean         396.179326    18.355428     2362.873522          2251.469724  
std         1708.620071    14.797316      422.057178           427.529246  
min            0.000000     1.000000     1575.000000          1286.000000  
25%          225.000000     1.000000     2017.000000          1894.000000  
50%          300.000000    25.000000     2298.000000          2192.000000  
75%          391.000000    30.000000     2676.000000          2533.000000  
max        49725.000000    71.000000     3398.000000          3387.000000

The rest of the variables seem to be in tolerable ranges, and now the data should be ready for regression models.

# Running ML models

We can now run machine learning models on the data.

In [67]:
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [68]:
# define y
y = final_data.cum_oil_365.values

# define X
X = final_data.drop(['cum_oil_365','api'],axis=1).values

In [69]:
X[:,1]

array([ 47.78861614,  47.90591376,  47.70375731, ...,  48.66200354,
        48.13889282,  47.76053993])

## Simple linear regression on all features

In [70]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=42)

In [71]:
reg = linear_model.LinearRegression()

reg.fit(X_train,y_train)

y_pred = reg.predict(X_test)

In [72]:
# R^2 value
reg.score(X_test, y_test)

0.30959730103584593

In [73]:
# RMSE value
np.sqrt(mean_squared_error(y_test, y_pred))

41552.083222097543

## Scaling values

In [74]:
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import scale

In [75]:
final_data.columns

Index(['api', 'cum_oil_365', 'TD', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'spud_timedelta', 'treatment_timedelta',
       'County_Name_BILLINGS', 'County_Name_BOTTINEAU', 'County_Name_BURKE',
       'County_Name_DIVIDE', 'County_Name_DUNN', 'County_Name_GOLDEN VALLEY',
       'County_Name_MCKENZIE', 'County_Name_MCLEAN', 'County_Name_MERCER',
       'County_Name_MOUNTRAIL', 'County_Name_RENVILLE', 'County_Name_STARK',
       'County_Name_WILLIAMS', 'Produced_Pools_BAKKEN',
       'Produced_Pools_BAKKEN/THREE FORKS', 'Produced_Pools_BIRDBEAR',
       'Produced_Pools_LODGEPOLE', 'Produced_Pools_MADISON',
       'Produced_Pools_ORDOVICIAN', 'Produced_Pools_RED RIVER',
       'Produced_Pools_SANISH', 'Produced_Pools_SPEARFISH',
       'Produced_Pools_SPEARFISH/MADISON', 'Produced_Pools_THREE FORKS',
       'Produced_Pools_TYLER'

In [76]:
data_to_scale = final_data.drop(['cum_oil_365','api'],axis=1)[['TD', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'spud_timedelta', 'treatment_timedelta']]
cat_values = final_data[['County_Name_BOTTINEAU', 'County_Name_BURKE', 'County_Name_DIVIDE',
       'County_Name_DUNN', 'County_Name_GOLDEN VALLEY', 'County_Name_MCKENZIE',
       'County_Name_MCLEAN', 'County_Name_MERCER', 'County_Name_MOUNTRAIL',
       'County_Name_RENVILLE', 'County_Name_STARK', 'County_Name_WILLIAMS',
       'Produced_Pools_BAKKEN/THREE FORKS', 'Produced_Pools_BIRDBEAR',
       'Produced_Pools_LODGEPOLE', 'Produced_Pools_MADISON',
       'Produced_Pools_ORDOVICIAN', 'Produced_Pools_RED RIVER',
       'Produced_Pools_SANISH', 'Produced_Pools_SPEARFISH',
       'Produced_Pools_SPEARFISH/MADISON', 'Produced_Pools_THREE FORKS',
       'Produced_Pools_TYLER', 'Wellbore_HORIZONTAL',
       'Wellbore_HORIZONTAL RE-ENTRY', 'Wellbore_VERTICAL']]

In [77]:
data_to_scale.head()

TD   Latitude   Longitude  bottom  fluid_bbl  fluid_bbl_per_ft  \
0  20548.0  47.788616 -102.860602   20548      81925                 8   
1  18748.0  47.905914 -102.751654   18080      63071                 9   
2  19990.0  47.703757 -102.481510   19677      22445                 2   
3  21125.0  47.788427 -103.398873   21125      40552                 4   
4  20790.0  47.978685 -102.681657   20519      71516                 7   

   ft_per_stage  lateral_length  max_treat_press  max_treat_rate  propp_lbs  \
0          9400            9400             8338            38.5    3285650   
1           221            6875             8762            32.3     363915   
2          8597            8597             9372            40.9    2807153   
3           323            9692             8206            25.6    1557517   
4           351            9133             8673            39.1    2625720   

   propp_lbs_per_ft  stages  spud_timedelta  treatment_timedelta  
0               349       1            2675                 2584  
1                52      31            2451                 2321  
2               326       1            2785                 2487  
3               160      30            1719                 1618  
4               287      26            2034                 2002

In [78]:
X_for_scale = data_to_scale.values
scaled_data = scale(X_for_scale)

In [79]:
pd.DataFrame(scaled_data).head()

0         1         2         3         4         5         6   \
0  0.455256 -0.487039  0.189397  0.490362  0.687421  0.041532  1.475186   
1 -0.280114 -0.233549  0.395909 -0.504112  0.236715  0.075156 -0.768077   
2  0.227291 -0.670426  0.907972  0.139395 -0.734451 -0.160208  1.278940   
3  0.690982 -0.487448 -0.830901  0.722862 -0.301603 -0.092961 -0.743149   
4  0.554122 -0.076285  0.528590  0.478676  0.438593  0.007909 -0.736306   

         7         8         9         10        11        12        13  \
0  0.360704  0.326399 -0.037891  0.087691 -0.027615 -1.172982  0.739602   
1 -0.891119  0.506714 -0.054133 -0.554302 -0.201455  0.854595  0.208821   
2 -0.037401  0.766130 -0.031604 -0.017449 -0.041077 -1.172982  1.000254   
3  0.505469  0.270263 -0.071684 -0.292032 -0.138240  0.787009 -1.525697   
4  0.228333  0.468865 -0.036320 -0.057316 -0.063905  0.516665 -0.779285   

         14  
0  0.777865  
1  0.162647  
2  0.550960  
3 -1.481832  
4 -0.583567

In [80]:
scaled_data = np.concatenate((scaled_data,cat_values.values),axis=1)

In [81]:
pd.DataFrame(scaled_data).head()

0         1         2         3         4         5         6   \
0  0.455256 -0.487039  0.189397  0.490362  0.687421  0.041532  1.475186   
1 -0.280114 -0.233549  0.395909 -0.504112  0.236715  0.075156 -0.768077   
2  0.227291 -0.670426  0.907972  0.139395 -0.734451 -0.160208  1.278940   
3  0.690982 -0.487448 -0.830901  0.722862 -0.301603 -0.092961 -0.743149   
4  0.554122 -0.076285  0.528590  0.478676  0.438593  0.007909 -0.736306   

         7         8         9  ...    31   32   33   34   35   36   37   38  \
0  0.360704  0.326399 -0.037891 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
1 -0.891119  0.506714 -0.054133 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
2 -0.037401  0.766130 -0.031604 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
3  0.505469  0.270263 -0.071684 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
4  0.228333  0.468865 -0.036320 ...   0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   

    39   40  
0  0.0  0.0  
1  0.0  0.0  
2  0.0  0.0  
3  0.0  0.0  
4  0.0  0.0  

[5 rows x 41 columns]

In [82]:
y.shape

(5582,)

In [83]:
scaled_train, scaled_test, y_train, y_test = train_test_split(scaled_data,y,test_size=0.3,random_state=42)

In [84]:
y_train.shape

(3907,)

In [85]:
reg_scale = linear_model.LinearRegression()

reg_scale.fit(scaled_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [86]:
y_pred = reg_scale.predict(scaled_test)
reg_scale.score(scaled_test, y_test)

0.30959730110467643

In [87]:
np.sqrt(mean_squared_error(y_test, y_pred))

41552.083220026245

## Random Forests

In [88]:
from sklearn.ensemble import RandomForestRegressor

In [89]:
rf = RandomForestRegressor(n_estimators=500,random_state=42)

In [90]:
rf.fit(scaled_train,y_train)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=500, n_jobs=1,
           oob_score=False, random_state=42, verbose=0, warm_start=False)

In [91]:
y_pred = rf.predict(scaled_test)
rf.score(scaled_test,y_test)

0.63708783181886386

In [92]:
np.sqrt(mean_squared_error(y_test, y_pred))

30126.05931779604

Clearly, random forests give a much better result than linear regression. This is as expected - some of the variables being fitted must have a non-linear relationship with cumulative oil production.

For the remaining part of the notebook, I stick to random forests as the regression tool.

## Adding in cross-validation

Here I compute the average R^2 and RMSE for taking different different folds, for random forests. 

In [93]:
from sklearn.model_selection import cross_val_score

In [94]:
cv_scores = cross_val_score(rf, scaled_data, y, cv=5)

In [95]:
cv_scores

array([ 0.60524218,  0.60048119,  0.5795629 ,  0.64373922,  0.66706015])

In [96]:
np.mean(cv_scores)

0.6192171289369407

In [97]:
cv_scores_mse = cross_val_score(rf, scaled_data, y, cv=5,scoring='mean_squared_error')

/usr/local/lib/python3.6/site-packages/sklearn/metrics/scorer.py:100: DeprecationWarning: Scoring method mean_squared_error was renamed to neg_mean_squared_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_weight)
/usr/local/lib/python3.6/site-packages/sklearn/metrics/scorer.py:100: DeprecationWarning: Scoring method mean_squared_error was renamed to neg_mean_squared_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_weight)
/usr/local/lib/python3.6/site-packages/sklearn/metrics/scorer.py:100: DeprecationWarning: Scoring method mean_squared_error was renamed to neg_mean_squared_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_weight)
/usr/local/lib/python3.6/site-packages/sklearn/metrics/scorer.py:100: DeprecationWarning: Scoring method mean_squared_error was renamed to neg_mean_squared_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_weight)
/usr/local/lib/python3.6/site-packages/sklearn/m

In [98]:
# so the mean squared error is reported as a negative value, and must be multiplied by -1
cv_scores_mse

array([ -8.72830811e+08,  -8.90980135e+08,  -8.93180019e+08,
        -9.85091046e+08,  -9.12418283e+08])

In [99]:
np.mean(np.sqrt(-1*cv_scores_mse)) # average RMSE value for all cv folds

30174.308090465609

## Feature engineering ideas:
1) Flag for maturity: Classify based on Produced_pools, and then bin them based on TD - this should indicate some measure of maturity. This should also somewhat capture engineering data like pressure, temperature, oil composition which are not available here.

2) Fracture intensity: max_treat_press should give idea on fracture intensity, for a given Produced_pool.

### Other attempts to reduce R^2/RMSE:
1) Take only horizontal wells

2) Center categorical variables

3) Put more estimators in the random forest

### Maturity flag

For guessing at maturity, it is not enough to bin by TD, but one must also group by Produced_pools, which gives a rough idea of maturity within a formation.

I have decided to create 10 different levels of maturity, corresponding to 10 different bins for the TD's of each produced pool. Therefore, this will result in the creation of a new categorical variable, 'Maturity', whose value ranges from 1-10, and is also dependent on formation, which is added as a prefix.

Some formations don't have enough data points. There will be only 1 maturity value for that formation.

In [100]:
final_data.head()

api  cum_oil_365       TD   Latitude   Longitude  bottom  \
0  33053032160000     262305.0  20548.0  47.788616 -102.860602   20548   
1  33053036170000     151827.0  18748.0  47.905914 -102.751654   18080   
2  33025010660000     168315.0  19990.0  47.703757 -102.481510   19677   
3  33053051550000      84342.0  21125.0  47.788427 -103.398873   21125   
4  33053043160000     104623.0  20790.0  47.978685 -102.681657   20519   

   fluid_bbl  fluid_bbl_per_ft  ft_per_stage  lateral_length  \
0      81925                 8          9400            9400   
1      63071                 9           221            6875   
2      22445                 2          8597            8597   
3      40552                 4           323            9692   
4      71516                 7           351            9133   

         ...          Produced_Pools_RED RIVER  Produced_Pools_SANISH  \
0        ...                                 0                      0   
1        ...                                 0                      0   
2        ...                                 0                      0   
3        ...                                 0                      0   
4        ...                                 0                      1   

   Produced_Pools_SPEARFISH  Produced_Pools_SPEARFISH/MADISON  \
0                         0                                 0   
1                         0                                 0   
2                         0                                 0   
3                         0                                 0   
4                         0                                 0   

   Produced_Pools_THREE FORKS  Produced_Pools_TYLER  Wellbore_DIRECTIONAL  \
0                           0                     0                     0   
1                           0                     0                     0   
2                           0                     0                     0   
3                           0                     0                     0   
4                           0                     0                     0   

   Wellbore_HORIZONTAL  Wellbore_HORIZONTAL RE-ENTRY  Wellbore_VERTICAL  
0                    1                             0                  0  
1                    1                             0                  0  
2                    1                             0                  0  
3                    1                             0                  0  
4                    1                             0                  0  

[5 rows x 46 columns]

The produced_pools column was deleted from final_data, and must be added back in to make the maturity flag.

In [101]:
# creating a new dataframe with_maturity to hold the final_data with maturity flag
with_maturity = final_data

In [102]:
prod_pools = well_indices[['api','Produced_Pools']]
prod_pools.head()

api Produced_Pools
0  33011013190000      RED RIVER
1  33025005960000         BAKKEN
2  33007015830000        MADISON
3  33007015840000        MADISON
4  33061005310000         BAKKEN

In [103]:
with_maturity = with_maturity.merge(prod_pools,how='inner',on='api')

In [104]:
with_maturity['Maturity'] = with_maturity['TD']

In [105]:
dict_pools = dict(list(with_maturity.groupby('Produced_Pools').Maturity))

In [106]:
length_dict = {key: len(value) for key, value in dict_pools.items()}

In [107]:
dict_pools['BIRDBEAR'].index

Int64Index([4584], dtype='int64')

In [108]:
grouped_data = with_maturity.groupby('Produced_Pools')

In [109]:
for pool in with_maturity.Produced_Pools.unique():
    cur_td = grouped_data.get_group(pool).TD
    if len(cur_td) < 10: # if the formation has less than 10 data points, use one bin for all of them
        n_bins = 1
    else:
        n_bins = 10
    cur_range = max(cur_td) - min(cur_td) + 1 # 1 is added to take care of cases where only one data point exists
    cur_min = min(cur_td)
    bin_size = cur_range/n_bins
    vals = grouped_data.get_group(pool).TD.map(lambda x: (x -cur_min)// bin_size + 1).astype('str')
    #vals = [pool + vals.astype('str')]
    with_maturity.loc[vals.index,'Maturity'] = vals

In [110]:
with_maturity.Maturity.unique()

array(['8.0', '7.0', '9.0', '5.0', '6.0', '10.0', '1.0', '2.0', '4.0',
       '3.0'], dtype=object)

In [112]:
with_maturity['Maturity'] = with_maturity['Maturity'] + '_' + with_maturity['Produced_Pools'] # add in formation names to the bins

In [113]:
with_maturity['Maturity'].head()

0    8.0_BAKKEN_BAKKEN
1    7.0_BAKKEN_BAKKEN
2    8.0_BAKKEN_BAKKEN
3    8.0_BAKKEN_BAKKEN
4    9.0_SANISH_SANISH
Name: Maturity, dtype: object

In [114]:
with_maturity['Maturity'].unique()

array(['8.0_BAKKEN_BAKKEN', '7.0_BAKKEN_BAKKEN', '9.0_SANISH_SANISH',
       '5.0_BAKKEN_BAKKEN', '9.0_BAKKEN_BAKKEN', '6.0_BAKKEN_BAKKEN',
       '10.0_SANISH_SANISH', '8.0_SANISH_SANISH',
       '1.0_SPEARFISH_SPEARFISH', '7.0_SPEARFISH_SPEARFISH',
       '10.0_SPEARFISH_SPEARFISH', '1.0_RED RIVER_RED RIVER',
       '1.0_SPEARFISH/MADISON_SPEARFISH/MADISON', '1.0_BAKKEN_BAKKEN',
       '10.0_MADISON_MADISON', '1.0_LODGEPOLE_LODGEPOLE',
       '10.0_BAKKEN_BAKKEN', '2.0_SANISH_SANISH',
       '1.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS', '6.0_SANISH_SANISH',
       '1.0_MADISON_MADISON', '1.0_SANISH_SANISH',
       '9.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS', '4.0_MADISON_MADISON',
       '3.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS',
       '8.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS', '7.0_SANISH_SANISH',
       '10.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS',
       '1.0_THREE FORKS_THREE FORKS',
       '2.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS', '5.0_SANISH_SANISH',
       '3.0_MADISON_

In [115]:
with_maturity[with_maturity['Produced_Pools']=='BAKKEN']['TD'].describe()

count     5440.000000
mean     19589.809007
std       2001.021643
min       1860.000000
25%      19115.000000
50%      20240.000000
75%      20785.500000
max      26908.000000
Name: TD, dtype: float64

In [116]:
with_maturity['Maturity'] = with_maturity['Maturity'].astype('category')

In [117]:
maturity_dummies = pd.get_dummies(with_maturity['Maturity'],prefix='mtrty',drop_first=True)

In [118]:
maturity_dummies.head()

mtrty_1.0_BAKKEN_BAKKEN  mtrty_1.0_BIRDBEAR_BIRDBEAR  \
0                        0                            0   
1                        0                            0   
2                        0                            0   
3                        0                            0   
4                        0                            0   

   mtrty_1.0_LODGEPOLE_LODGEPOLE  mtrty_1.0_MADISON_MADISON  \
0                              0                          0   
1                              0                          0   
2                              0                          0   
3                              0                          0   
4                              0                          0   

   mtrty_1.0_RED RIVER_RED RIVER  mtrty_1.0_SANISH_SANISH  \
0                              0                        0   
1                              0                        0   
2                              0                        0   
3                              0                        0   
4                              0                        0   

   mtrty_1.0_SPEARFISH/MADISON_SPEARFISH/MADISON  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   mtrty_1.0_SPEARFISH_SPEARFISH  mtrty_1.0_THREE FORKS_THREE FORKS  \
0                              0                                  0   
1                              0                                  0   
2                              0                                  0   
3                              0                                  0   
4                              0                                  0   

   mtrty_10.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS           ...             \
0                                                 0           ...              
1                                                 0           ...              
2                                                 0           ...              
3                                                 0           ...              
4                                                 0           ...              

   mtrty_7.0_BAKKEN_BAKKEN  mtrty_7.0_SANISH_SANISH  \
0                        0                        0   
1                        1                        0   
2                        0                        0   
3                        0                        0   
4                        0                        0   

   mtrty_7.0_SPEARFISH_SPEARFISH  \
0                              0   
1                              0   
2                              0   
3                              0   
4                              0   

   mtrty_8.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS  mtrty_8.0_BAKKEN_BAKKEN  \
0                                                0                        1   
1                                                0                        0   
2                                                0                        1   
3                                                0                        1   
4                                                0                        0   

   mtrty_8.0_SANISH_SANISH  mtrty_8.0_SPEARFISH_SPEARFISH  \
0                        0                              0   
1                        0                              0   
2                        0                              0   
3                        0                              0   
4                        0                              0   

   mtrty_9.0_BAKKEN/THREE FORKS_BAKKEN/THREE FORKS  mtrty_9.0_BAKKEN_BAKKEN  \
0                                                0                        0   
1                                                0                        0   
2                                                0    

In [119]:
scaled_maturity = np.concatenate((scaled_data,maturity_dummies.values),axis=1)

In [120]:
maturity_train, maturity_test, y_mat_train, y_mat_test = train_test_split(scaled_maturity,y,test_size=0.3,random_state=42)

In [121]:
rf_maturity = RandomForestRegressor(n_estimators=500,random_state=100)
rf_maturity.fit(maturity_train,y_mat_train)
y_mat_pred = rf_maturity.predict(maturity_test)
rf_maturity.score(maturity_test,y_mat_test)

0.63681864825334267

In [122]:
np.sqrt(mean_squared_error(y_mat_test, y_mat_pred))

30137.229977769432

In [123]:
cv_scores_maturity = cross_val_score(rf_maturity, scaled_maturity, y, cv=5)
np.mean(cv_scores_maturity)

0.61771652291041657

Clearly, adding in the maturity flag hardly had any effect on the R^2 values.

### Dropping vertical wells

In [124]:
only_horizontal = with_maturity.drop(labels=['Wellbore_HORIZONTAL','Wellbore_HORIZONTAL RE-ENTRY', 'Wellbore_VERTICAL'],axis=1)

In [125]:
well_bore_info = well_indices[['api','Wellbore']]

In [126]:
only_horizontal = only_horizontal.merge(well_bore_info,how='inner',on='api')

In [127]:
only_horizontal.Wellbore.unique()

array(['HORIZONTAL', 'VERTICAL', 'DIRECTIONAL', 'HORIZONTAL RE-ENTRY'], dtype=object)

In [128]:
only_horizontal = only_horizontal[only_horizontal['Wellbore'] == 'HORIZONTAL']

In [129]:
only_horizontal.columns

Index(['api', 'cum_oil_365', 'TD', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'spud_timedelta', 'treatment_timedelta',
       'County_Name_BILLINGS', 'County_Name_BOTTINEAU', 'County_Name_BURKE',
       'County_Name_DIVIDE', 'County_Name_DUNN', 'County_Name_GOLDEN VALLEY',
       'County_Name_MCKENZIE', 'County_Name_MCLEAN', 'County_Name_MERCER',
       'County_Name_MOUNTRAIL', 'County_Name_RENVILLE', 'County_Name_STARK',
       'County_Name_WILLIAMS', 'Produced_Pools_BAKKEN',
       'Produced_Pools_BAKKEN/THREE FORKS', 'Produced_Pools_BIRDBEAR',
       'Produced_Pools_LODGEPOLE', 'Produced_Pools_MADISON',
       'Produced_Pools_ORDOVICIAN', 'Produced_Pools_RED RIVER',
       'Produced_Pools_SANISH', 'Produced_Pools_SPEARFISH',
       'Produced_Pools_SPEARFISH/MADISON', 'Produced_Pools_THREE FORKS',
       'Produced_Pools_TYLER'

In [130]:
horizontal_to_scale = only_horizontal[['TD', 'Latitude', 'Longitude', 'bottom',
       'fluid_bbl', 'fluid_bbl_per_ft', 'ft_per_stage', 'lateral_length',
       'max_treat_press', 'max_treat_rate', 'propp_lbs', 'propp_lbs_per_ft',
       'stages', 'spud_timedelta', 'treatment_timedelta']]
horizontal_cat = only_horizontal[['County_Name_BOTTINEAU', 'County_Name_BURKE', 'County_Name_DIVIDE',
       'County_Name_DUNN', 'County_Name_GOLDEN VALLEY', 'County_Name_MCKENZIE',
       'County_Name_MCLEAN', 'County_Name_MERCER', 'County_Name_MOUNTRAIL',
       'County_Name_RENVILLE', 'County_Name_STARK', 'County_Name_WILLIAMS',
       'Produced_Pools_BAKKEN/THREE FORKS', 'Produced_Pools_BIRDBEAR',
       'Produced_Pools_LODGEPOLE', 'Produced_Pools_MADISON',
       'Produced_Pools_ORDOVICIAN', 'Produced_Pools_RED RIVER',
       'Produced_Pools_SANISH', 'Produced_Pools_SPEARFISH',
       'Produced_Pools_SPEARFISH/MADISON', 'Produced_Pools_THREE FORKS',
       'Produced_Pools_TYLER']]

In [131]:
scaled_horizontal = scale(horizontal_to_scale.values)

In [132]:
scaled_horizontal = np.concatenate((scaled_horizontal,horizontal_cat.values),axis=1)

In [133]:
y_horizontal = only_horizontal['cum_oil_365'].values

In [134]:
hor_train, hor_test, y_hor_train, y_hor_test = train_test_split(scaled_horizontal,y_horizontal,test_size=0.3,random_state=42)

In [135]:
hor_test.shape

(1667, 38)

In [136]:
rf_hor = RandomForestRegressor(n_estimators=500,random_state=100)
rf_hor.fit(hor_train,y_hor_train)
y_hor_pred = rf_hor.predict(hor_test)
rf_hor.score(hor_test,y_hor_test)

0.59169219778670168

In [140]:
np.sqrt(mean_squared_error(y_hor_test,y_hor_pred))

31319.43382440904

In [137]:
only_hor_cv = cross_val_score(rf_hor, scaled_horizontal, y_horizontal, cv=5)

In [138]:
only_hor_cv

array([ 0.60398354,  0.5911527 ,  0.57935467,  0.63823268,  0.66638664])

In [139]:
np.mean(only_hor_cv)

0.61582204524973849

Keeping only the horizontal wells results in some cv-folds having high R^2 values, but others are low, and don't seem to improve results.

### Adding in factor for fracture intensity


In [141]:
# creating a new dataframe with_maturity to hold the final_data with maturity flag
with_frac_intensity = final_data

In [142]:
with_frac_intensity = with_frac_intensity.merge(prod_pools,how='inner',on='api')

In [145]:
with_frac_intensity['Frac_intensity'] = with_frac_intensity['max_treat_press']

In [146]:
grouped_data = with_frac_intensity.groupby('Produced_Pools')

In [147]:
for pool in with_frac_intensity.Produced_Pools.unique():
    cur_press = grouped_data.get_group(pool).max_treat_press
    if len(cur_press) < 10: # if the formation has less than 10 data points, use one bin for all of them
        n_bins = 1
    else:
        n_bins = 10
    cur_range = max(cur_press) - min(cur_press) + 1 # 1 is added to take care of cases where only one data point exists
    cur_min = min(cur_press)
    bin_size = cur_range/n_bins
    vals = grouped_data.get_group(pool).max_treat_press.map(lambda x: (x -cur_min)// bin_size + 1).astype('str')
    #vals = [pool + vals.astype('str')]
    with_frac_intensity.loc[vals.index,'Frac_intensity'] = vals

In [148]:
with_frac_intensity.Frac_intensity.unique()

array(['5.0', '7.0', '3.0', '4.0', '1.0', '6.0', '2.0', '8.0', '10.0',
       '9.0'], dtype=object)

In [149]:
with_frac_intensity['Frac_intensity'] = with_frac_intensity['Frac_intensity'] + '_' + with_frac_intensity['Produced_Pools'] # add in formation names to the bins

In [150]:
with_frac_intensity['Frac_intensity'].head()

0    5.0_BAKKEN
1    5.0_BAKKEN
2    5.0_BAKKEN
3    5.0_BAKKEN
4    7.0_SANISH
Name: Frac_intensity, dtype: object

In [152]:
with_frac_intensity['Frac_intensity'].unique()

array(['5.0_BAKKEN', '7.0_SANISH', '3.0_BAKKEN', '4.0_BAKKEN',
       '1.0_BAKKEN', '6.0_BAKKEN', '6.0_SANISH', '2.0_SPEARFISH',
       '1.0_SPEARFISH', '5.0_SANISH', '2.0_BAKKEN', '8.0_SANISH',
       '1.0_RED RIVER', '1.0_SPEARFISH/MADISON', '4.0_SANISH',
       '1.0_MADISON', '3.0_SPEARFISH', '10.0_SPEARFISH', '1.0_LODGEPOLE',
       '7.0_BAKKEN', '1.0_BAKKEN/THREE FORKS', '10.0_SANISH',
       '5.0_MADISON', '10.0_BAKKEN/THREE FORKS', '3.0_MADISON',
       '4.0_MADISON', '9.0_SPEARFISH', '8.0_BAKKEN/THREE FORKS',
       '4.0_SPEARFISH', '9.0_BAKKEN/THREE FORKS', '10.0_MADISON',
       '6.0_BAKKEN/THREE FORKS', '1.0_SANISH', '1.0_THREE FORKS',
       '7.0_BAKKEN/THREE FORKS', '5.0_SPEARFISH', '10.0_BAKKEN',
       '9.0_SANISH', '1.0_BIRDBEAR', '7.0_SPEARFISH'], dtype=object)

In [153]:
with_frac_intensity['Frac_intensity'] = with_frac_intensity['Frac_intensity'].astype('category')

In [154]:
frac_dummies = pd.get_dummies(with_frac_intensity['Frac_intensity'],prefix='fracInten')

In [155]:
frac_dummies.head()

fracInten_1.0_BAKKEN  fracInten_1.0_BAKKEN/THREE FORKS  \
0                     0                                 0   
1                     0                                 0   
2                     0                                 0   
3                     0                                 0   
4                     0                                 0   

   fracInten_1.0_BIRDBEAR  fracInten_1.0_LODGEPOLE  fracInten_1.0_MADISON  \
0                       0                        0                      0   
1                       0                        0                      0   
2                       0                        0                      0   
3                       0                        0                      0   
4                       0                        0                      0   

   fracInten_1.0_RED RIVER  fracInten_1.0_SANISH  fracInten_1.0_SPEARFISH  \
0                        0                     0                        0   
1                        0                     0                        0   
2                        0                     0                        0   
3                        0                     0                        0   
4                        0                     0                        0   

   fracInten_1.0_SPEARFISH/MADISON  fracInten_1.0_THREE FORKS  \
0                                0                          0   
1                                0                          0   
2                                0                          0   
3                                0                          0   
4                                0                          0   

            ...             fracInten_6.0_SANISH  fracInten_7.0_BAKKEN  \
0           ...                                0                     0   
1           ...                                0                     0   
2           ...                                0                     0   
3           ...                                0                     0   
4           ...                                0                     0   

   fracInten_7.0_BAKKEN/THREE FORKS  fracInten_7.0_SANISH  \
0                                 0                     0   
1                                 0                     0   
2                                 0                     0   
3                                 0                     0   
4                                 0                     1   

   fracInten_7.0_SPEARFISH  fracInten_8.0_BAKKEN/THREE FORKS  \
0                        0                                 0   
1                        0                                 0   
2                        0                                 0   
3                        0                                 0   
4                        0                                 0   

   fracInten_8.0_SANISH  fracInten_9.0_BAKKEN/THREE FORKS  \
0                     0                                 0   
1                     0                                 0   
2                     0                                 0   
3                     0                                 0   
4                     0                                 0   

   fracInten_9.0_SANISH  fracInten_9.0_SPEARFISH  
0                     0                        0  
1                     0                        0  
2                     0                        0  
3                     0                        0  
4                     0                        0  

[5 rows x 40 columns]

In [156]:
scaled_frac_inten = np.concatenate((scaled_data,frac_dummies.values),axis=1)

In [157]:
frac_train, frac_test, y_frac_train, y_frac_test = train_test_split(scaled_frac_inten,y,test_size=0.3,random_state=42)

In [158]:
rf_frac = RandomForestRegressor(n_estimators=500,random_state=100)
rf_frac.fit(frac_train,y_frac_train)
y_frac_pred = rf_frac.predict(frac_test)
rf_frac.score(frac_test,y_frac_test)

0.63785028564324953

In [159]:
np.sqrt(mean_squared_error(y_frac_test, y_frac_pred))

30094.396276011015

In [161]:
cv_scores_frac = cross_val_score(rf_frac, scaled_frac_inten, y, cv=5)
np.mean(cv_scores_frac)

0.61830808829011497

### Rescale categorical variables

The reason I tried this was because the documentation for scaling says the variables should be scaled as well as centered. The categorical variables were not centered.

In [163]:
scaled_data_2 = scale(X_for_scale)
rep = {0:-1}
new_cat_values = cat_values.replace(rep)

In [164]:
new_cat_values.head()

County_Name_BOTTINEAU  County_Name_BURKE  County_Name_DIVIDE  \
0                     -1                 -1                  -1   
1                     -1                 -1                  -1   
2                     -1                 -1                  -1   
3                     -1                 -1                  -1   
4                     -1                 -1                  -1   

   County_Name_DUNN  County_Name_GOLDEN VALLEY  County_Name_MCKENZIE  \
0                -1                         -1                     1   
1                -1                         -1                     1   
2                 1                         -1                    -1   
3                -1                         -1                     1   
4                -1                         -1                     1   

   County_Name_MCLEAN  County_Name_MERCER  County_Name_MOUNTRAIL  \
0                  -1                  -1                     -1   
1                  -1                  -1                     -1   
2                  -1                  -1                     -1   
3                  -1                  -1                     -1   
4                  -1                  -1                     -1   

   County_Name_RENVILLE        ...          Produced_Pools_ORDOVICIAN  \
0                    -1        ...                                 -1   
1                    -1        ...                                 -1   
2                    -1        ...                                 -1   
3                    -1        ...                                 -1   
4                    -1        ...                                 -1   

   Produced_Pools_RED RIVER  Produced_Pools_SANISH  Produced_Pools_SPEARFISH  \
0                        -1                     -1                        -1   
1                        -1                     -1                        -1   
2                        -1                     -1                        -1   
3                        -1                     -1                        -1   
4                        -1                      1                        -1   

   Produced_Pools_SPEARFISH/MADISON  Produced_Pools_THREE FORKS  \
0                                -1                          -1   
1                                -1                          -1   
2                                -1                          -1   
3                                -1                          -1   
4                                -1                          -1   

   Produced_Pools_TYLER  Wellbore_HORIZONTAL  Wellbore_HORIZONTAL RE-ENTRY  \
0                    -1                    1                            -1   
1                    -1                    1                            -1   
2                    -1                    1                            -1   
3                    -1                    1                            -1   
4                    -1                    1                            -1   

   Wellbore_VERTICAL  
0                 -1  
1                 -1  
2                 -1  
3                 -1  
4                 -1  

[5 rows x 26 columns]

In [165]:
scaled_data_2 = np.concatenate((scaled_data_2,new_cat_values.values),axis=1)

In [166]:
scaled_train_2, scaled_test_2, y_train_2, y_test_2 = train_test_split(scaled_data_2,y,test_size=0.3,random_state=42)

In [167]:
rf_scaled_cat = RandomForestRegressor(n_estimators=500,random_state=100)
rf_scaled_cat.fit(scaled_train_2,y_train_2)
y_pred_2 = rf_scaled_cat.predict(scaled_test_2)
rf_scaled_cat.score(scaled_test_2,y_test_2)

0.63568207397306287

In [168]:
cv_scores_2 = cross_val_score(rf_scaled_cat, scaled_data_2, y, cv=5)
np.mean(cv_scores_2)

0.61758863501200101

In [169]:
cv_scores_2

array([ 0.60363963,  0.59772154,  0.57691758,  0.6430631 ,  0.66660134])

### Adding more estimators to the random forest

In [170]:
rf_scaled_cat = RandomForestRegressor(n_estimators=1000,random_state=100)
rf_scaled_cat.fit(scaled_train_2,y_train_2)
y_pred_2 = rf_scaled_cat.predict(scaled_test_2)
rf_scaled_cat.score(scaled_test_2,y_test_2)

0.63654946593483275

In [171]:
np.sqrt(mean_squared_error(y_test_2,y_pred_2))

30148.396447041894

# Conclusion

The best R^2 values observed was around 0.66, with an RMSE of 30000 bbls.

The random forest model performed significantly better than the linear regression, but the feature engineering ideas did not have a significant effect on the R^2 and/or the RMSE values.

It is possible that doing the data cleaning more intelligently would generate a better model. The data cleaning I did was mostly based on what values of a variable should be reasonable. This may not necessarily be true.